platform/ios/src/MGLMapView.mm

```diff
@@ -680,6 +696,45 @@ - (void)reachabilityChanged:(NSNotification *)notification
     _isWaitingForRedundantReachableNotification = NO;
 }
 
+// This is similar to macOS' `restorableStateKeyPaths` used for state restoration
+// Support for `UIStateRestoring` is still TBD: see https://github.com/mapbox/mapbox-gl-native/issues/3660
++ (NSArray *)mglRestorableStateKeyPaths {
+    return @[@"camera", @"debugMask", @"styleURL"];
+}
+
+- (id)valueAtTerminationForKey:(NSString*)key ofClass:(Class)classType {
+    id value = self.propertyValuesAtTermination[key];
+    
+    if (![value isKindOfClass:classType]) {
+        return nil;
+    }
+    
+    return value;
+}
+
+- (void)destroyCoreObjects {
+    // Record the current state
+    self.propertyValuesAtTermination = [NSMutableDictionary dictionary];
+    
+    for (NSString *property in [self.class mglRestorableStateKeyPaths]) {
+        self.propertyValuesAtTermination[property] = [self valueForKey:property];
+    }
+    
+    // Tear down C++ objects, insuring worker threads correctly terminate.
+    // Because of how _mbglMap is constructed, we need to destroy it first.
+    delete _mbglMap;
+    _mbglMap = nullptr;
```

Comment:
`_mbglMap` is declared as an `mbgl::Map &`. A C++ reference can’t ever become null, because [a reference isn’t a pointer](https://stackoverflow.com/a/4364586/4585461). The Objective-C++ compiler apparently lets you get away with setting the reference to null, but using the reference is undefined behavior – even checking if the reference evaluates to `false`.

Reply:
`_mbglMap` is still a `mbgl::Map *` - it's just the accessor that I changed to a C++ reference. This was to indicate that it should never be null, and if it is then a crash would be valid (hence raising the exception in the getter).

I'm in two minds about whether the getter should return a reference or pointer. Returning the pointer does indicate that care needs to be taken; OTOH we should have already taken care :)